### ISSUE TITLE
Support updating jobs / scheduled workflows

### ISSUE THREAD

**Issue by discordianfish** [kubeflow/pipelines#3789]
Currently there is no way to update an job / scheduled workflow via the pipeline API. The only way is deleting it first, then recreating it. But if there is already a workflow running, it has the scheduledWorkflow as ownerReference set, so the workflow (and all pods it spawned) get deleted. That prevents us from automatically updating the jobs from our workflow repository in CI.

/kind feature
/area backend
/area sdk

**Comment by discordianfish** [comment]
Dunno about long term plans or general design vision behind the api, but if the right way of adding this is adding a `PATCH /apis/v1beta1/jobs/{id}` route to swagger and add the code to `apply` the changes to the underlying scheduledWorkflow, I can submit a PR for that.

**Comment by jonasdebeukelaer** [comment]
Hey I've submitted a pr which should make this easy to add via the api https://github.com/kubeflow/pipelines/pull/3788

It's a bit more complete than just patching so metadata and params can be updated too. I have ui changes ready as well, once this gets through. 

issue: https://github.com/kubeflow/pipelines/issues/1180

**Comment by Bobgy** [comment]
@discordianfish @jonasdebeukelaer 
Instead of editing a Job directly, I'm thinking about an alternative:
```
1. disable the job (existing runs won't be removed)
2. clone the job and change the pipeline/version and scheduled interval start time
```

This approach works with existing features, and leaves a track record of what jobs were scheduled with what pipeline. What do you think?

When we update a Job, what should its behavior for existing runs be? what about auto-backfill? Do people want to know edit history of this Job? I think there's a lot of scope for confusion.

Maybe the disable and clone approach can be better even in the long term.

Also
/assign @jingzhang36 @Bobgy

**Comment by discordianfish** [comment]
@Bobgy This is reasonable but from my POV an implementation detail. We need to update jobs. Whether this internally is disabling the old, enabling a new doesn't matter much to me. Another option would be dropping jobs altogether. In it's current from its pretty much useless tbh. If it wouldn't have been there, I would just have sticked to my k8s cronjobs submitting the workflow.

My 2 cents on meta: https://github.com/kubeflow/pipelines/pull/3788 has been open for over two months with no clear indication of which direction to go and now this pretty specific design proposal. This makes it seem like this project isn't really open source and decisions are my within Google without really involving the contributors. I'm worried that will hurt the project.

**Comment by jonasdebeukelaer** [comment]
Sorry I've been pretty quiet for a while. @discordianfish I'd agree that jobs don't seem to add any value (at least in my limited experience working with them), and all it does is add a bunch of complexity, especially here when trying to do a humble update. the only thing missing would be a description - could you include this with a cronjob somehow? As a custom metadata?

If jobs _are_ useful, then @Bobgy maybe disabling and creating new jobs is probably cleaner if you do anticipate history being a request down the line. Seems reasonable to me.

**Comment by Bobgy** [comment]
@discordianfish I agree these are implementation details, it was more of a discussion in the context of where we want to go for #3788. You are welcome to stick to k8s cronjob if this isn't useful for you right now.

We'll document Job's featureset and readiness soon, so I hope this won't be a confusion e.g. after spending so much time trying it out.

For #3788, https://github.com/kubeflow/pipelines/pull/3788#issuecomment-651497592 was the blocker that @jonasdebeukelaer didn't respond to (no blame, it's totally understandable if you have other focuses). If our scheduledworkflow controller doesn't handle CRD instance update well, there's no value just adding the api method.
I don't think that's a problem with governance.

Thanks for bringing up `CronJob`, it's currently Beta. When it matures, we can consider swapping out KFP job's underlying implementation to it.

**Comment by Bobgy** [comment]
I synced with @jingzhang36 today, our conclusion for Job is:
1. We'll provide these features as is, if cloning a Job doesn't work for you, please use other alternatives like kubernetes `CronJob`.
2. We may work on a refactoring that swaps out underlying `ScheduledWorkflow` implementation with kubernetes `CronJob`, that will be best for long term maintenance. We don't want to maintain our own `ScheduledWorkflowController` that is a feature subset of `CronJob`. Using `CronJob` as implementation details allow first party integration like UI management, metadata for pipeline runs for CronJob, while keeping maintenance cost low for the controller.

FYI, you might not understand the problem in https://github.com/kubeflow/pipelines/pull/3788#issuecomment-651497592. Note that `ScheduledWorkflow` is a CRD introduced by KFP. For its cron feature, we have our own controller implementation in https://github.com/kubeflow/pipelines/tree/master/backend/src/crd/controller/scheduledworkflow. The controller was implemented without taking trigger/pipeline update behavior into consideration. So your PR for just adding an API method for patching the `ScheduledWorkflow` CRD instance may not be enough. We need a deep dive into the controller and figure out what is preferred behavior for various kinds of edge cases that can happen.

Because cost for maintaining `ScheduledWorkflow` controller is high, we are inclined to swap it with `CronJob` resource instead, so that we can focus on application logic.

We don't have an ETA for the refactoring right now, because `Job` is not our priority, we'll triage based on future user requirements.

**Comment by discordianfish** [comment]
> For #3788, #3788 (comment) was the blocker that @jonasdebeukelaer didn't respond to (no blame, it's totally understandable if you have other focuses). If our scheduledworkflow controller doesn't handle CRD instance update well, there's no value just adding the api method.

Fair enough. So you are okay with merging the PR if the ScheduledWorkflowController handles this as expected?
If that's the case, I can spend time creating a cluster to validate that. But I'm worried that, like in the past, this PR doesn't end going anywhere because of decisions I'm not aware of that lead into a different direction.


> I don't think that's a problem with governance.
Well, this probably isn't the best place to discuss governance. I'll try to bring it up on some community meetings in the future.

Just to be clear, I don't expect anyone to implement features/changes I'm asking for. But I need to know which changes will be accepted and which won't to be able to contribute. And raising such questions on the mailinglist didn't really work either.

**Comment by Bobgy** [comment]
@discordianfish Thanks for explaining the confusions you are getting! I agree it would be a painful experiencing spending a lot of pain writing a PR, but not seeing it being treated. I think for anything that requires an interface change, we'd better have an agreement on the design first.

I've seen some other projects always require an enhancement proposal before actually working on a new feature, that's probably a way to resolve your confusion. We can agree on the feature request and design first, before starting to write production ready PR (proof of concept PRs still welcomed at your own interest).

I can try to surface the problem in our team internally if we can create such a structure formally. Do you think that will mitigate the issue?

**Comment by Bobgy** [comment]
> Fair enough. So you are okay with merging the PR if the ScheduledWorkflowController handles this as expected?
If that's the case, I can spend time creating a cluster to validate that. But I'm worried that, like in the past, this PR doesn't end going anywhere because of decisions I'm not aware of that lead into a different direction.

Yes, if we can verify it's working as intended. I can help you keep the PR reviewed and approved.

Especially, I'll list some non exhaustive cases I'm interested in:
* changing trigger type
* changing begin date and end date (e.g. expanding begin date, expanding end date ...)
* how does the above work with catchup: true/catchup: false
* changing pipeline/pipeline version

and we need to consider whether behavior for these cases are appropriate or not

It would be best if you can read https://github.com/kubeflow/pipelines/tree/master/backend/src/crd/controller/scheduledworkflow and giving some code links proving your idea.

====
So I really think this is a lot of work, if you feel that's important for you, please go ahead, but please do quick investigations.
and if it turns out that we need to do implementation changes for scheduledworkflow controller to support these use cases, I don't think we'll approve big PRs. Scheduler is not core value of KFP and we'll want to use a mature scheduler as implementation details later.

Back to your original user request, I don't clearly see why it's important having an edit API. Why it doesn't work with "disable + clone"? Do you use UI or SDK as main client? Can we add client side helpers to make this workflow easier without touching scheduledworkflow controller?

**Comment by discordianfish** [comment]
> I've seen some other projects always require an enhancement proposal before actually working on a new feature, that's probably a way to resolve your confusion.

Something along those lines would help. Doesn't even have to be that formal. Personally I'd prefer just discussing implementation in an issue like this one. An alternative would be discussing on the mailinglist. But neither of these usually get attention from maintainers which IMO is the root of the problem. Willing contributors are blocked by not getting feedback on their suggestion, so you have to resort to 'trial and error'-PRs.

I'm a maintainer of https://github.com/prometheus/node_exporter where we have a lightweight process of labeling issues as 'accepted' if it's something we want to support or 'require feedback' if we still need more input from users to decide. For more fundamental changes we ask to raise it on the mailinglist. But whatever works best for you guys. The only thing I'd wish for is some semi-guaranteed way to get feedback on feature requests like this one.

Now back to the original request: I think I misunderstood what you meant initially. So to update a job, you propose we:

- Disable the old job (and not delete it as I'm doing right now)
- Create a new job
- Eventually clean up the old job

I suppose that would work for us, but it's non-trivial to know when it's safe to remove the old job. But if there is a way to version the job, that should work for us. I'll explore that option.

For context: We're using jsonnet to generate our argo workflows (that was built long before we migrated to kf) in CI and using the kfp python lib to upload them.

**Comment by Bobgy** [comment]
Thanks for the suggestions for community support. I'll surface that for internal discussion.

Yes, and `Eventually clean up the old job` is not a necessary requirement. If we want to support this workflow better, we can let UI hide disabled Jobs by default and use an option to see them.

> For context: We're using jsonnet to generate our argo workflows (that was built long before we migrated to kf) in CI and using the kfp python lib to upload them.

I see, if you are not using UI, do you know how to use KFP API filters? for example, you can filter by a substring match in Job name and only those enabled.
This is an example: https://github.com/kubeflow/pipelines/pull/4087/files#diff-1f65a44a2187458da0f00e0ebbe730e5R322-R330.

I think filters will enable your workflow to use Jobs themselves as versions, you can set up a naming convention that a Job name prefix corresponds to a Job identity and in CI, you can disable the active Job for that and clone a new one with the same name prefix.

**Comment by Bobgy** [comment]
also FYI, our documentation for generated client just got sooo much better than before: https://kubeflow-pipelines.readthedocs.io/en/latest/source/kfp.server_api.html#kfp_server_api.api.job_service_api.JobServiceApi.list_jobs

**Comment by discordianfish** [comment]
@bobgy Okay my plan (essentially) is to redo our CI scripts:

- upload new job and set version=git-sha in description
- deactivate all jobs with version!=git-sha in the description

But it doesn't look like the filter allows for IS_NOT_SUBSTRING. Am I missing something?

**Comment by Bobgy** [comment]
> @bobgy Okay my plan (essentially) is to redo our CI scripts:
> 
> - upload new job and set version=git-sha in description
> - deactivate all jobs with version!=git-sha in the description
> 
> But it doesn't look like the filter allows for IS_NOT_SUBSTRING. Am I missing something?

@discordianfish Can you give each logical job a special job name prefix?

So that you can list all jobs with that prefix and use python to filter descriptions without the git tag you are talking about.

**Comment by Bobgy** [comment]
Closing because we will not support update   for now.

**Comment by Bobgy** [comment]
/reopen

There were further discussions in https://github.com/kubeflow/pipelines/issues/4752.

/cc @StefanoFioravanzo @NikeNano 

I think we can restart discussion here.

**Comment by thesuperzapper** [comment]
@chensun @james-jwu @zijianjoy I think this issue is critical for the wider adoption of Kubeflow Pipelines. Right now it is effectively impossible to make a git-ops system that manages kubeflow pipelines.

## Problem

Say we want to make a git-ops system to manage which pipelines are scheduled on a Kubeflow cluster. 
I have a working prototype of such a system, but it's very sub-optimal because:

 1. Scheduled Runs (AKA Jobs) can't be updated
 2. Deleting a Scheduled Run will delete any currently running Argo Workflows (because they have an owner-reference set) 

This means any solution will be inherently brittle because it relies on "disabling" old Scheduled Runs, the user must:

- create a new Scheduled Run for any change 
- disable all previous "versions" of this Schedule Run
- either have an ever-expanding history of these versions, or clean them up once they get too old 

__This solution is NOT perfect:__

1. During the "change over" period, the `max_concurrency` will not be respected:
     - If the pipeline was already running, multiple versions could run at the same time.
     - The only other option is to unceremoniously delete all old versions (which would kill workflows mid-run).
3.  Having many old (but disabled) Scheduled Runs is dangerous and messy:
     - It relies on consistent naming to determine which Scheduled Runs are the same, for example: `"{job_name}-{updated-timestamp}"`
     - It massively clutters the UI, making it hard to know which runs are enabled.

## Solution

The best solution is to create a [PATCH on the `/apis/v2beta1/recurringruns/{recurring_run_id}` API](https://www.kubeflow.org/docs/components/pipelines/v2/reference/api/kubeflow-pipeline-api-spec/#tag-RecurringRunService) for updating existing Scheduled Runs.

I don't think the semantics of this are particularly complex as there are only two cases:

1. Updating the pipeline definition:
     - In this case, we would be updating the definition for FUTURE runs, leaving any currently in-flight runs to finish.
     - This would be the default anyway, because pipelines are not executed by KFP but Argo Workflows, so are static and out of our control after they start.
2. Updating the trigger schedule:
     - In this case, we can simply use the new schedule going forwards.

## Related Issues

There is also a related long-term request for us to make the "source of truth" for Scheduled Runs the `ScheduledRun` CRD.

- https://github.com/kubeflow/pipelines/issues/6001

This CRD already exists, but purely as a back-end thing managed by the KFP API, if users update the CRD, everything breaks, as it will become out-of-sync with the KFP MySQL database.


/reopen

**Comment by thesuperzapper** [comment]
Hey everyone, I have released a reference repository that demonstrates using GitOps to manage the schedules and definitions of Kubeflow Pipelines.

### You can find it at [`deployKF/kubeflow-pipelines-gitops`](https://github.com/deployKF/kubeflow-pipelines-gitops)

It has limitations caused by Kubeflow Pipeline not allowing updates to "recurring runs" (only deleting and recreating), but I think the solution it uses will work for most production use-cases.

**Comment by rvolchek** [comment]
🏁 flagged for review
